libs/execution/src/lib/blocks/block-execution-util.ts

```diff
@@ -0,0 +1,107 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import {
+  BlockDefinition,
+  collectParents,
+} from '@jvalue/jayvee-language-server';
+
+import { ExecutionContext } from '../execution-context';
+import { Logger } from '../logger';
+import { IOTypeImplementation, NONE } from '../types';
+
+// eslint-disable-next-line import/no-cycle
+import { createBlockExecutor } from './block-executor-registry';
+import * as R from './execution-result';
+
+export interface ExecutionOrderItem {
+  block: BlockDefinition;
+  value: IOTypeImplementation | null;
+}
+
+export async function executeBlocks(
+  executionContext: ExecutionContext,
+  executionOrder: ExecutionOrderItem[],
+  initialInputValue: IOTypeImplementation | undefined = undefined,
+): Promise<R.Result<ExecutionOrderItem[]>> {
+  let isFirstblock = true;
+
+  for (const blockData of executionOrder) {
+    const block = blockData.block;
+    const parentData = collectParents(block).map((parent) =>
+      executionOrder.find((blockData) => parent === blockData.block),
+    );
+    let inputValue =
+      parentData[0]?.value === undefined ? NONE : parentData[0]?.value;
+
+    if (
+      isFirstblock &&
+      inputValue === NONE &&
+      initialInputValue !== undefined
+    ) {
+      inputValue = initialInputValue;
+    }
```

Comment:
In context of the comment above on the `initialInputValue` - maybe it makes sense to use an explanatory variable for the condition to express the code's intent more clearly.